clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java

```diff
@@ -406,15 +407,13 @@ private static DefaultRecord readFrom(ByteBuffer buffer,
             String headerKey = Utils.utf8(buffer, headerKeySize);
             buffer.position(buffer.position() + headerKeySize);
 
-            ByteBuffer headerValue = null;
+            byte[] headerValue = null;
             int headerValueSize = ByteUtils.readVarint(buffer);
             if (headerValueSize >= 0) {
-                headerValue = buffer.slice();
-                headerValue.limit(headerValueSize);
+                headerValue = Utils.toArray(buffer, headerValueSize);
```

Comment:
So here the only thing we do is create the header object which is string, byte[].  As per kip. Should not the string is memory copied already.

As such we can make this a buffer but a line or two down when we hand over to the header object it would be a byte[].